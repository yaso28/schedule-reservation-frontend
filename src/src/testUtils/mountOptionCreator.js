import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
