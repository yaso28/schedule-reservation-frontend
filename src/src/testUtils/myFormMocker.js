import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
