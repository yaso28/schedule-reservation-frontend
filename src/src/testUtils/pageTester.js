import { mount } from '@vue/test-utils'
import MyPage from '@/components/containers/MyPage'

class PageTester {
  async mount (pageComponent, route, mountOption) {
    await mountOption.router.push(route)
    const pageWrapper = mount(MyPage, mountOption)
    return pageWrapper.findComponent(pageComponent)
  }

  assertTitleTag () {
    expect(document.title).toMatchSnapshot()
  }

  assertMetaTags () {
    expect(document.querySelectorAll('meta[data-vue-meta]')).toMatchSnapshot()
  }
}

export default new PageTester()
