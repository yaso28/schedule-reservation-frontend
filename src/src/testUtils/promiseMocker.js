import _ from 'lodash'

const createSetTimeoutPromise = getResult => new Promise((resolve, reject) => {
  setTimeout(() => getResult(resolve, reject), 1000)
})
const createSwitchPromise = (value, error, switchObj) => createSetTimeoutPromise(
  (resolve, reject) => switchObj.isSuccess ? resolve(value) : reject(error)
)
const createResolvedPromise = value => createSetTimeoutPromise(resolve => resolve(value))
const createRejectedPromise = error => createSetTimeoutPromise((resolve, reject) => reject(error))

const createMock = createPromise => jest.fn().mockImplementationOnce(createPromise)
const createMockForCases = (caseList, createPromiseForCurrentCase) => jest.fn((...args) => {
  for (let i = 0; i < caseList.length; i++) {
    const currentCase = caseList[i]
    if (_.isEqual(args, currentCase.args)) {
      return createPromiseForCurrentCase(currentCase)
    }
  }
  throw new Error('createMockForCases: no args matched')
})

class PromiseMocker {
  init () {
    jest.useFakeTimers()
  }

  clear () {
    jest.clearAllTimers()
  }

  flushPendings () {
    jest.runOnlyPendingTimers()
  }

  createSwitch () {
    return { isSuccess: true }
  }

  setSwitch (switchObj, isSuccess) {
    switchObj.isSuccess = isSuccess
  }

  mock (value, error, switchObj) {
    return createMock(() => createSwitchPromise(value, error, switchObj))
  }

  mockResolved (value) {
    return createMock(() => createResolvedPromise(value))
  }

  mockRejected (error) {
    return createMock(() => createRejectedPromise(error))
  }

  mockForCases (caseList) {
    return createMockForCases(caseList, currentCase => createSwitchPromise(currentCase.value, currentCase.error, currentCase.switch))
  }

  mockResolvedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createResolvedPromise(currentCase.value))
  }

  mockRejectedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createRejectedPromise(currentCase.error))
  }
}

export default new PromiseMocker()
