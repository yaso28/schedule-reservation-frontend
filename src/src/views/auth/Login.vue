<template>
  <div>
    <slot name="h1" />

    <CForm @submit.prevent="login">
      <CCard>
        <CCardBody>
          <MyText
            :label="$t('user.id')"
            :form="form"
            field="email"
          />
          <MyText
            :label="$t('user.password')"
            :form="form"
            field="password"
            type="password"
          />
        </CCardBody>
        <CCardFooter>
          <CButton
            type="submit"
            color="primary"
          >
            {{ $t("action.login") }}
          </CButton>
        </CCardFooter>
      </CCard>
    </CForm>
  </div>
</template>

<script>
export default {
  name: 'AuthLogin',
  components: {
    MyText: require('@/components/inputs/MyText').default
  },
  inject: ['apiService'],
  inheritAttrs: false,
  data () {
    return {
      form: this.apiService.makeForm({
        email: '',
        password: ''
      })
    }
  },
  methods: {
    async login () {
      try {
        this.$store.commit('processing/beginPost')
        this.$store.commit('alert/clear')
        const result = await this.apiService.login(this.form)
        this.$store.dispatch('auth/set', result)
        this.$router.push({ name: 'home' })
      } catch (e) {
        if (!this.apiService.isHandledError(e)) {
          throw e
        }
      } finally {
        this.$store.commit('processing/endPost')
      }
    }
  }
}
</script>
