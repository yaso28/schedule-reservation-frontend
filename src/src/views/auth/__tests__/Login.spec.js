import Login from '../Login'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import pageTester from '@/testUtils/pageTester'
import promiseMocker from '@/testUtils/promiseMocker'
import storeModules from '@/store/modules'
import myFormMocker from '@/testUtils/myFormMocker'

jest.mock('@/store/modules')
jest.mock('@/models/MyForm')

function getFormWrapper (wrapper) {
  return wrapper.find('form')
}

describe('Login', () => {
  let mountOption
  let router
  let apiService
  beforeEach(() => {
    promiseMocker.init()
    mountOption = mountOptionCreator.create(storeModules)
    router = mountOption.router
    apiService = mountOption.provide.apiService
  })

  afterEach(() => {
    promiseMocker.clear()
    jest.resetAllMocks()
  })

  describe('マウント', () => {
    let form
    let wrapper
    beforeEach(async () => {
      form = myFormMocker.createMock({
        email: '',
        password: ''
      })
      apiService.makeForm = jest.fn().mockReturnValueOnce(form)
      wrapper = await pageTester.mount(Login, { name: 'login' }, mountOption)
    })

    test('titleタグ', () => {
      pageTester.assertTitleTag()
    })

    test('metaタグ', () => {
      pageTester.assertMetaTags()
    })

    test('ページ全体表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    describe('フォーム送信', () => {
      let loginSwitch
      let loginResult
      let loginError
      beforeEach(() => {
        loginSwitch = promiseMocker.createSwitch()
        loginResult = require('@/testUtils/dummyApiData/auth/login.json')
        loginError = new Error()
        apiService.login = promiseMocker.mock(
          loginResult,
          loginError,
          loginSwitch
        )
        getFormWrapper(wrapper).trigger('submit')
      })

      test('Processingカバー表示開始', () => {
        expect(storeModules.processing.mutations.beginPost).toHaveBeenCalled()
      })

      test('ログイン処理を呼出', () => {
        expect(apiService.login).toHaveBeenCalledWith(form)
      })

      test('Processingカバー表示中', () => {
        expect(
          storeModules.processing.mutations.endPost
        ).not.toHaveBeenCalled()
      })

      describe('ログイン：エラー', () => {
        beforeEach(() => {
          promiseMocker.setSwitch(loginSwitch, false)
          apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
          promiseMocker.flushPendings()
        })

        test('エラー処理', () => {
          expect(apiService.isHandledError).toHaveBeenCalledWith(loginError)
        })

        test('Processingカバー表示終了', () => {
          expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
        })
      })

      describe('ログイン：成功', () => {
        beforeEach(() => {
          promiseMocker.setSwitch(loginSwitch, true)
          router.push = jest.fn()
          promiseMocker.flushPendings()
        })

        test('認証情報セット', () => {
          expect(storeModules.auth.actions.set.mock.calls[0][1]).toBe(
            loginResult
          )
        })

        test('ホームページへ転送', () => {
          expect(router.push).toHaveBeenCalledWith({ name: 'home' })
        })

        test('Processingカバー表示終了', () => {
          expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
        })
      })
    })
  })
})
