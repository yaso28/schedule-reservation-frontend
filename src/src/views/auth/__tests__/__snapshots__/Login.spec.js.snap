// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login マウント metaタグ 1`] = `
NodeList [
  <meta
    content="noindex"
    data-vmid="robots"
    data-vue-meta="2"
    name="robots"
  />,
]
`;

exports[`Login マウント titleタグ 1`] = `"やそスポーツクラブ | ログイン"`;

exports[`Login マウント ページ全体表示 1`] = `
<div>
  <h1>
    ログイン
  </h1>
   
  <form
    class=""
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="email"
          >
             ID/メールアドレス 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <input
              class="form-control"
              id="email"
              type="text"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="password"
          >
             パスワード 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <input
              class="form-control"
              id="password"
              type="password"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
      </div>
       
      <footer
        class="card-footer"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          
          ログイン
        
        </button>
      </footer>
    </div>
  </form>
</div>
`;
